server/src/main/java/io/crate/common/collections/RefCountedItem.java

```diff
@@ -21,33 +21,56 @@
 
 package io.crate.common.collections;
 
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.ArrayList;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
 
 public class RefCountedItem<T> implements AutoCloseable {
 
-    private final T item;
-    private final Runnable onClose;
-    private final AtomicInteger refs = new AtomicInteger(1);
+    private final Function<String, T> itemFactory;
+    private final Consumer<T> closeItem;
+    private final ArrayList<String> sources = new ArrayList<>();
```

Comment:
do we need to clear it on close?

Reply:
Wasn't sure about that. It's not really required - the life-time of the RefCountedItem is for a single query, so it should never grow large and it may be interesting for debugging to track all acquisition calls.